gapis/perfetto/android/trace.go

```diff
@@ -172,6 +172,16 @@ func Start(ctx context.Context, d adb.Device, a *android.ActivityAction, opts *s
 		}
 	}
 
+	// Enforce the AGI to launch the perfetto data producer, this will kill the
+	// running perfetto data producer and re-launch it if it's already running.
+	// And hence, any sequential update after perfetto data sources are queried
+	// will likely lead to undefined behaviour.
+	// If there's an error happen, likely the perfetto data producer or
+	// the device has issues, log the error, ignore it and keep proceeding.
```

Comment:
The current implementation of EnsurePerfettoProducerLaunched seems to never return an error.

It feels like EnsurePerfettoProducerLaunched may silence some errors, but should still propagate errors from launchPerfettoProducer(). And here we should fail if an error is reported.

What do you think?

Reply:
I am familiar with golang, but I think since `launchPerfettoProducer` is within its own function we can't propagate errors from it.

Reply:
You can propagate the error outside of the crash.Go lambda via a channel. It's not obvious, I've seen that approach in used in the `gapir/client` package.

Here, this should do the trick:

```
--- a/gapidapk/gapidapk.go
+++ b/gapidapk/gapidapk.go
@@ -228,19 +228,22 @@ func ensurePerfettoProducerLaunched(ctx context.Context, d adb.Device) error {
 
        startSignal, startFunc := task.NewSignal()
        startFunc = task.Once(startFunc)
+       launchErr := make(chan error)
        crash.Go(func() {
                err := launchPerfettoProducer(ctx, d, driver, startFunc)
                if err != nil {
                        log.E(ctx, "[ensurePerfettoProducerLaunched] error: %v", err)
                }
+               launchErr <- err
 
                // Ensure the start signal is fired on failure/immediate return.
                startFunc(ctx)
        })
        startSignal.Wait(ctx)
+       err = <-launchErr
        // TODO(b/148420473): Data sources are queried before gpu.counters is registered.
        time.Sleep(1 * time.Second)
-       return nil
+       return err
 }
```

Reply:
that would block `EnsurePerfettoProducerLaunched` from returning, right? `launchPerfettoProducer` may not return immediately.

Reply:
If I understand the task framework correctly, you're already waiting for launchPerfettoProducer() to terminate with `startSignal.Wait(ctx)`? And so, by the time `startSignal.Wait(ctx)` returns, you're guaranteed to have something in launchErr to read?

Reply:
no, `startSignal.Wait(ctx)` either waits for `launchPerfettoProducer` to signal where in that case, it won't return so that means the `crash.Go` is still stuck at `launchPerfettoProducer`, or `launchPerfettoProducer` to fail and the `startFunc(ctx)` is called. So `err = <- launchErr` will stuck if everything goes well.

I suppose we should just pass `launchErr` in probably.

Reply:
Oh yeah, my naive approach here would deadlock, sorry! Passing launchErr to `launchPerfettoProducer` sounds good.